pkg/queue/health/probe.go

```diff
@@ -63,18 +95,85 @@ var transport = func() *http.Transport {
 	return t
 }()
 
-// HTTPProbe checks that HTTP connection can be established to the address.
-func HTTPProbe(config HTTPProbeConfigOptions) error {
+func getURL(config HTTPProbeConfigOptions) url.URL {
+	url := url.URL{
+		Scheme: string(config.Scheme),
+		Host:   net.JoinHostPort(config.Host, config.Port.String()),
+		Path:   config.Path,
+	}
+	return url
+}
+
+// http2UpgradeProbe checks that an HTTP with HTTP2 upgrade request
+// connection can be understood by the address.
+func http2UpgradeProbe(config HTTPProbeConfigOptions) (*bool, bool, error) {
+
 	httpClient := &http.Client{
 		Transport: transport,
 		Timeout:   config.Timeout,
 	}
+	url := getURL(config)
+	req, err := http.NewRequest(http.MethodOptions, url.String(), nil)
+	if err != nil {
+		return nil, false, fmt.Errorf("error constructing probe request %w", err)
+	}
 
-	url := url.URL{
-		Scheme: string(config.Scheme),
-		Host:   net.JoinHostPort(config.Host, config.Port.String()),
-		Path:   config.Path,
+	// An upgrade will need to have at least these 3 headers.
+	req.Header.Add("Connection", "Upgrade, HTTP2-Settings")
+	req.Header.Add("Upgrade", "h2c")
+	req.Header.Add("HTTP2-Settings", "")
+
+	req.Header.Add(network.UserAgentKey, network.KubeProbeUAPrefix+config.KubeMajor+"/"+config.KubeMinor)
+
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return nil, false, err
+	}
+	defer res.Body.Close()
+
+	// It is possible that the container might not be ready yet. We
+	if res.StatusCode >= 400 {
+		return nil, false, fmt.Errorf("HTTP probe did not respond Ready, got status code: %d", res.StatusCode)
+	}
+
+	isHTTP2 := false
+	isHTTP1AndReady := false
+	if IsHTTPProbeReady(res) {
+		isHTTP1AndReady = true
+	} else if res.StatusCode == http.StatusSwitchingProtocols && res.Header.Get("Upgrade") == "h2c" {
+		isHTTP2 = true
+	}
+
+	return &isHTTP2, isHTTP1AndReady, nil
+}
+
+// HTTPProbe checks that HTTP connection can be established to the address.
+func HTTPProbe(ctx context.Context, config HTTPProbeConfigOptions) error {
+	cfg := apicfg.FromContextOrDefaults(ctx)
+	autoDetect := cfg.Features.AutoDetectHTTP2
+	if autoDetect == apicfg.Disabled {
+		ishttp := false
+		config.IsHTTP2 = &ishttp
```

Comment:
reading through Im not sure I totally understand why IsHTTP2 needs to be a bool-pointer rather than just a straight boolean. Is it possible to replace with just a normal boolean?

Reply:
Thanks @julz! You're right; the original change was a bit different, and this is a (probably unnecessary) artifact of that. Let me refactor it.

Reply:
I have refactored this class, instead of using IsHTTP bool, I have changed it to MaxProtoMajor, making '0' a much safer "unknown" state. Let me know if this sounds better! Also, thank you for the in-depth discussion in Slack, I learned a few things today :)!

Reply:
Refactored into using MaxProtoVersion instead of IsHTTP2 bool, I think it's easier and safer to reason about the "unknown"state. Thanks for the extended discussion on Slack, I learned a few things today!